core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -777,8 +783,70 @@ protected void convertSelectImpl(
     }
   }
 
+
+
+
+  /**
+   * Having translated 'SELECT ... FROM ... [GROUP BY ...] [HAVING ...]',
+   * handles the Qualify Expression.
+   *
+   * We expect that, while handling the conversion of the select List, the QUALIFY condition was
+   * added as the last element of the selectList. Therefore, this function uses that value to create
+   * a filter and projection (see
+   * https://docs.snowflake.com/en/sql-reference/constructs/qualify.html)
+   *
+   * @param bb               Blackboard
+   */
+  private void handleQualifyClause(
+      Blackboard bb) {
+    RelNode rel = bb.root;
+    // If we had a qualify clause, we know it's the last item in the select list,
+    // as we place it there.
+
+    final List<RelDataTypeField> fields = rel.getRowType().getFieldList();
+    final RexNode filterVal = RexInputRef.of(fields.size() - 1, fields);
+
+    // I'm not entirly certain why we do this, but the normal path that handles WHERE does this,
+    // So I'm going to leave it.
+    final RexNode filterValWithoutNullability = RexUtil.removeNullabilityCast(typeFactory,
```

Comment:
I think you do this for the dereference since its nullable. You may need to do this elsewhere.